documentation/projects/proposals/search_relevancy_sandbox/20230518-implementation_plan_staging_index_rapid_iteration.md

```diff
@@ -0,0 +1,357 @@
+   will be replaced entirely (see `override_config` parameter below).
+4. `index_suffix`: (Optional) The name suffix of the new index to create. This
+   will be a string, and will be used to name the index in Elasticsearch of the
+   form `{media_type}-{index_suffix}`. If not provided, the suffix will be a
+   timestamp of the form `YYYYMMDDHHMMSS`.
+5. `source_index`: (Optional) The existing index on Elasticsearch to use as the
+   basis for the new index. If not provided, the index aliased to `media_type`
+   will be used (e.g. `image` for the `image` media type).
+6. `override_config`: (Optional) A boolean value which can be toggled to replace
+   the existing index configuration entirely with the new configuration. If
+   `True`, the `index_config` parameter will be used as the entire
+   configuration. If `False`, the `index_config` parameter will be merged with
+   the existing configuration. Defaults to `False`.
+7. `query`: (Optional) An
+   [Elasticsearch query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html)
+   to use to filter the documents to be copied to the new index. If not
+   provided, all documents will be copied. See
+   [`create_and_populate_filtered_index` on the ingestion server](https://github.com/WordPress/openverse/blob/3fcce5ade2165955db5bbcb4f679257b3260547b/ingestion_server/ingestion_server/indexer.py#L529)
+   for an example of how this is used.
+
+### DAG
+
+Once the parameters are provided, the DAG will execute the following steps:
+
+1. Create the index configuration. If `override_config` is supplied, the
+   `index_config` parameter will be used as the entire configuration. If not,
+   the `index_config` parameter will be merged with the configuration generated
+   by `es_mapping::index_settings`. See [Merging policy](#merging-policy) for
+   how this merge will be performed.
+2. Create the new index with the configuration generated in step 1. The index
+   name will either be a combination of the `media_type` and `index_suffix`
+   parameters (`{media_type}-{index_suffix}`, e.g. `image-my-special-suffix`),
+   or the `media_type` and a timestamp (`{media_type}-{timestamp}`, e.g.
+   `image-20200102030405`). This will be done using the
+   [`ElasticsearchPythonHook`](https://airflow.apache.org/docs/apache-airflow-providers-elasticsearch/stable/_api/airflow/providers/elasticsearch/hooks/elasticsearch/index.html#airflow.providers.elasticsearch.hooks.elasticsearch.ElasticsearchPythonHook)
+   ([example from the existing ingestion server code](https://github.com/WordPress/openverse/blob/3fcce5ade2165955db5bbcb4f679257b3260547b/ingestion_server/ingestion_server/indexer.py#L518-L521)).
+3. Initiate a reindex using the `source_index` as the source
+   ([example from the ingestion server](https://github.com/WordPress/openverse/blob/3fcce5ade2165955db5bbcb4f679257b3260547b/ingestion_server/ingestion_server/indexer.py#L525-L547)).
```

Comment:
I have a question regarding this step. While thinking about the plan for #1987, I wondered if these plans are not more closely related and dependent on each other. Since the DAG will be using the [`_reindex` API of ES](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/docs-reindex.html), it assumes _there exists at least one index_ already, presumably the result of a data refresh. Am I correct? If so, it's worth pointing this out explicitly in the dependencies. The first time I read it, I overlooked it.

Reply:
I agree the DAG would be indifferent to which index is iterated on –although it will be quite relevant to the person triggering– still, it requires _one_ index to derive from, as I said before, due to the use of the Reindex API of ES. Which will be this very base index now that the staging DB is constantly being replaced? Perhaps this is what I should cover in the IP for the issue mentioned above but I having a hard time figuring out this part 😅

Reply:
Ahhh, I see what you mean! Implicit to my thinking was that one of those indices, be it the proportional-by-provider index (my pick) or the production-data-volume, would be aliased to `image` (or `audio` depending on the media type). Right now we do have those aliases, and they point to indices that were hand-made: 
![image](https://github.com/WordPress/openverse/assets/10214785/9bf86848-d2bb-4a56-8ab3-a8dda060f157)

I do think that's a piece to be hashed out as part of the final IP for this project (i.e. how to create those indices and which should be used as the "primary" alias). I'm realizing now though that "Staging Elasticsearch reindex DAGs for both potential index types" is maybe not the best name for that effort, potentially "Staging Elasticsearch index creation DAGs for both potential index types" is better? This is maybe even a discussion to have on that issue instead 😅